### ISSUE TITLE
Alarm Stop & Snooze Notification Action Buttons

### ISSUE THREAD

**Issue by Toxicann** [gdelataillade/alarm#58]
**Alarm plugin version**
alarm 1.1.5 

**Describe the bug**
I have created custom action Buttons for snooze & stop using flutter local notications package. The stop & snooze function works if it is triggered from other parts of file. But It throws  " Unhandled Exception: LateInitializationError: Field 'prefs' has not been initialized." error when those functions(stop & snooze) are accessed from onreceivedNotifications listener of the flutter local notifications package.

**To Reproduce**
*Steps to reproduce the behavior:*
1. Create a custom stop & snooze notification
2. listen for notification
  await flutterLocalNotificationsPlugin.initialize(
    initializationSettings,
    onDidReceiveNotificationResponse: onTapForeground,
    onDidReceiveBackgroundNotificationResponse: onTapBackground,
  );

@pragma('vm:entry-point')
Future<void> onTapBackground(NotificationResponse message) async {
  log('tapped a background message ${message.payload}');
  await Alarm.stop(message.id!);
}

The await Alarm.stop throws this error:
 [ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: LateInitializationError: Field 'prefs' has not been initialized.

**Expected behavior**
*Alarm should have stopped*

**Comment by gdelataillade** [comment]
Hi @Toxicann,

Thanks for your interest in the plugin !

This is not the first time is see this error. I found this solution in StackOverflow:
https://stackoverflow.com/questions/71065727/lateinitializationerror-sharedpreferences

Make sure you call WidgetsFlutterBinding.ensureInitialized() and then Alarm.init in your main function. This is how I did in the example app:

```
Future<void> main() async {
  WidgetsFlutterBinding.ensureInitialized();
  SystemChrome.setPreferredOrientations([DeviceOrientation.portraitUp]);

  await Alarm.init(showDebugLogs: true);

  runApp(const MaterialApp(home: ExampleAlarmHomeScreen()));
}
```

Let me know if it works.

**Comment by Toxicann** [comment]
This is not working for me as you can see i already have  WidgetsFlutterBinding.ensureInitialized(); called

![image](https://github.com/gdelataillade/alarm/assets/31942766/96a86164-7056-4763-bd41-3c3560ad1f4c)

**Comment by gdelataillade** [comment]
The issue comes from `Alarm.init` being called within the `onTapBackground` function, which operates in a different isolate. As isolates don't share memory in Dart, SharedPreferences isn't accessible here. This becomes problematic when `Alarm.stop` tries to delete the saved alarm from SharedPreferences, leading to an error.

I'll think about it and aim to come up with a solution or workaround. If you have any ideas or suggestions, I'd love to hear them. Your input is always valuable!

**Comment by Toxicann** [comment]
Is there an alternative way in which i can achieve this stop, Snooze function form notifications?

**Comment by gdelataillade** [comment]
Hey @Toxicann,

I'm doing some research. Just to be clear, what you want is to be able to stop your alarm without opening your app, just by using the notification Stop button ?

**Comment by Toxicann** [comment]
Yes, I wanted to make a stop button & snooze just by using notification buttons. 
I couldn't find a way around that.

**Comment by gdelataillade** [comment]
OK. Sorry I don't have an alternative for you at the moment. There are some limitations:

Method channels cannot be used in the callback (or any function running in a background isolate) because they rely on the Flutter engine, which isn't available in a background isolate.

I don't give up, I believe I will be able to find a workaround, but for now I don't know how sorry.

**Comment by sagar1garg** [comment]
Hi is there any progress on this? @gdelataillade

**Comment by gdelataillade** [comment]
Hi @sagar1garg

I would love to work on that but for now there are issues with much higher priority. I'll keep you updated.

**Comment by gdelataillade** [comment]
Hi @sagar1garg @Toxicann 

I released version [4.0.0-dev.4](https://pub.dev/packages/alarm/versions/4.0.0-dev.4) with notification action button STOP. Please let me know if it works good for you. Be careful of breaking changes. Also, it's still a development version but I will soon release stable version `4.0.0`.
